File: day01.py
"""Solution do day 1 of Advent of Code"""

from get_input import get_input, line_parser

import itertools


def part1(lines):
    for a, b in itertools.combinations(lines, 2):
        if a + b == 2020:
            return a * b


def part2(lines):
    for a, b, c in itertools.combinations(lines, 3):
        if a + b + c == 2020:
            return a * b * c
    raise NotImplementedError


if __name__ == "__main__":
    LINES = line_parser(get_input(day=1, year=2020))
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day01.rs
use common::get_input;
use itertools::Itertools;

type Input = Vec<isize>;

fn find_size(nums: &Vec<isize>, target: isize, len: usize) -> Option<isize> {
    for combo in nums.into_iter().combinations(len) {
        if combo.clone().into_iter().sum::<isize>() == target {
            return Some(combo.into_iter().product::<isize>())
        }
    }
    None
}

fn part1(input: &Input) -> isize {
    find_size(input, 2020, 2).unwrap()
}

fn part2(input: &Input) -> isize {
    find_size(input, 2020, 3).unwrap()
}

fn parse(text: String) -> Input {
    text.trim()
        .split('\n')
        .map(|n| n.parse::<isize>().unwrap())
        .collect()
}

fn main() {
    let input = parse(get_input(01, 2020));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
